### PR TITLE
Add new extension points for keyboard input and speech pause

### PR DESCRIPTION
**Pull request by ctoth**
Add new extension points for raw keyboard events and speech pause state

This commit introduces two new extension points:

1. inputCore.decide_handleRawKey
- Called before any NVDA processing of raw keyboard events
- Allows add-ons to intercept and optionally block keyboard events
- Provides full context: vkCode, scanCode, extended flag, and press/release state
- Implemented in both keyDown and keyUp event handlers
- Returns True to allow normal processing, False to block the key

2. speech.extensions.speechPaused
- Notifies when speech is paused or resumed 
- Provides boolean 'switch' parameter (True=paused, False=resumed)
- Added corresponding unit tests to verify functionality
- Integrated into existing pauseSpeech() function

Technical Details:
- Added documentation for both extension points in developerGuide.md
- Updated speech/__init__.py to expose the new speechPaused extension point
- Added test case in test_speech.py to verify extension point behavior
- Maintains backwards compatibility with existing extensions

These additions enable add-ons to:
- Implement advanced keyboard interception/modification
- React to speech pause state changes

### Link to issue number:
N/A - New feature addition for extensibility

### Summary of the issue:
NVDA needed additional extension points to allow add-ons to:
1. Intercept and control raw keyboard events before NVDA processing
2. Monitor and react to speech pause state changes

### Description of user facing changes
- No direct user-facing changes
- Enables add-on developers to create more sophisticated keyboard handling and speech feedback features
- All changes are API-level additions that maintain backwards compatibility

### Description of development approach
1. Raw Keyboard Extension Point:
   - Added `decide_handleRawKey` extension point in inputCore.py
   - Integrated into both keyDown and keyUp event handlers in keyboardHandler.py
   - Full keyboard event context provided (vkCode, scanCode, extended, pressed)
   - Boolean return value controls event propagation

2. Speech Pause Extension Point:
   - Added `speechPaused` extension point in speech/extensions.py
   - Integrated into existing pauseSpeech() function
   - Provides pause state through boolean parameter

3. Documentation:
   - Added entries in developerGuide.md
   - Updated relevant module documentation
   - Added changelog entry

### Testing strategy:
1. Unit Tests:
   - Added test cases in test_speech.py for speechPaused extension point
   - Tests verify extension point notification on pause/resume
   - Tests ensure correct parameter passing

2. Manual Testing:
   - Verified keyboard extension point with test add-on
   - Confirmed speech pause notifications work as expected

### Known issues with pull request:
None identified - straightforward API additions with full test coverage

### Code Review Checklist:

- [x] Documentation:
  - Change log entry added in changes.md
  - Developer documentation updated in developerGuide.md
  - Technical documentation added in module docstrings
  - No GUI changes, so no context help needed

- [x] Testing:
  - Unit tests added for speechPaused
  - Manual testing performed with test add-on
  - System tests not required (API-level changes)

- [x] UX of all users considered:
  - No direct user impact
  - Enables better add-on support for all output methods
  - No localization impact (API only)

- [x] API is compatible with existing add-ons
  - New extension points are additive only
  - No breaking changes to existing APIs
  - Follows established extension point patterns

- [x] Security precautions taken:
  - Discussed raw keystroke decider security implications with NV Access.

@coderabbitai summary

### CODE REVIEW

File: source/inputCore.py [hunk 1]
```diff
@@ -452,6 +452,17 @@ def __eq__(self, other: Any) -> bool:
...
 
+decide_handleRawKey = extensionPoints.Decider()
+"""
+Notifies when a raw keyboard event is received, before any NVDA processing.
+Handlers can decide whether the key should be processed by NVDA and/or passed to the OS.
+@param vkCode: The virtual key code
+@param scanCode: The scan code
+@param extended: Whether this is an extended key
+@param pressed: Whether this is a key press or release
+@return: True to allow normal processing, False to block the key
```
Comment (seanbudd) [inline source/inputCore.py:463]: can you add type information to these like other extensionPoints. We don't do this for most functions, but we don't have a way of adding type hints directly with extensionPoints yet.

File: source/inputCore.py [hunk 2]
```diff
@@ -452,6 +452,17 @@ def __eq__(self, other: Any) -> bool:
 		return NotImplemented
 
 
+decide_handleRawKey = extensionPoints.Decider()
```
Comment (seanbudd) [inline source/inputCore.py:455]: could unit tests be added for this?

File: user_docs/en/changes.md [hunk 3]
```diff
@@ -99,6 +99,10 @@ Add-ons will need to be re-tested and have their manifest updated.
 * Removed the requirement to indent function parameter lists by two tabs from NVDA's Coding Standards, to be compatible with modern automatic linting. (#17126, @XLTechie)
 * Added the [VS Code workspace configuration for NVDA](https://nvaccess.org/nvaccess/vscode-nvda) as a git submodule. (#17003)
 * In the `brailleTables` module, a `getDefaultTableForCurrentLang` function has been added (#17222, @nvdaes)
+* Added the following extension points:
```
Comment (seanbudd) [inline user_docs/en/changes.md:102]:
```suggestion
* Added the following extension points (#17428):
```

File: user_docs/en/changes.md [hunk 4]
```diff
@@ -99,6 +99,10 @@ Add-ons will need to be re-tested and have their manifest updated.
...
+  * inputCore.decide_handleRawKey: called on each keypress
+  * speech.extensions.speechPaused: Called when speech is paused or unpaused
```
Comment (seanbudd) [inline user_docs/en/changes.md:104]:
```suggestion
  * `inputCore.decide_handleRawKey`: called on each keypress
  * `speech.extensions.speechPaused`: Called when speech is paused or unpaused
```

File: source/speech/extensions.py [hunk 5]
```diff
@@ -22,6 +22,14 @@
 Handlers are called without arguments.
 """
 
+speechPaused = Action()
```
Comment (seanbudd) [inline source/speech/extensions.py:25]:
```suggestion
post_speechPaused = Action()
```

From codingStandards: https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/codingStandards.md#identifier-names

**Review by seanbudd — COMMENTED** [review]
Thanks @ctoth - this seems almost ready

File: source/inputCore.py [hunk 6]
```diff
@@ -452,6 +452,22 @@ def __eq__(self, other: Any) -> bool:
...
+@param vkCode: The virtual key code
+@type vkCode: int
+@param scanCode: The scan code
+@type scanCode: int
+@param extended: Whether this is an extended key
+@type extended: bool
+@param pressed: Whether this is a key press or release
+@type pressed: bool
+@return: True to allow normal processing, False to block the key
+@rtype: bool
```
Comment (seanbudd) [inline source/inputCore.py:468]: Sphinx style docs please

```suggestion
:param vkCode: The virtual key code
:type vkCode: int
:param scanCode: The scan code
:type scanCode: int
:param extended: Whether this is an extended key
:type extended: bool
:param pressed: Whether this is a key press or release
:type pressed: bool
:return: True to allow normal processing, False to block the key
:rtype: bool
```

File: source/speech/extensions.py [hunk 7]
```diff
@@ -22,6 +22,14 @@
 Handlers are called without arguments.
 """
 
+post_speechPaused = Action()
+"""
+Notifies when speech is paused.
+
+@param switch: True if speech is paused, False if speech is resumed.
+@type switch: bool
```
Comment (seanbudd) [inline source/speech/extensions.py:30]:
```suggestion
:param switch: True if speech is paused, False if speech is resumed.
:type switch: bool
```

**Comment by seanbudd** [comment]
Does this close #14520?

File: source/inputCore.py [hunk 8]
```diff
@@ -452,6 +452,22 @@ def __eq__(self, other: Any) -> bool:
 		return NotImplemented
 
 
+decide_handleRawKey = extensionPoints.Decider()
```
Comment (CyrilleB79) [inline source/inputCore.py:455]: Would it not be better to indicate the signature of the handler in the docstring and remove each individual param type description?

File: user_docs/en/changes.md [hunk 9]
```diff
@@ -104,6 +104,9 @@ Add-ons will need to be re-tested and have their manifest updated.
 * Added the [VS Code workspace configuration for NVDA](https://nvaccess.org/nvaccess/vscode-nvda) as a git submodule. (#17003)
 * In the `brailleTables` module, a `getDefaultTableForCurrentLang` function has been added (#17222, @nvdaes)
 * Retrieving the `labeledBy` property now works for UIA elements supporting the corresponding `LabeledBy` UIA property. (#17442, @michaelweghorn)
+* Added the following extension points (#17428):
```
Comment (CyrilleB79) [inline user_docs/en/changes.md:107]: This is optional, but maybe you are interested in adding your GitHub contact ID:

```suggestion
* Added the following extension points (#17428, @ctoth):
```

**Comment by CyrilleB79** [comment]
Is this PR triggered by the need to remove monkey patching in NVDA remote?
Or have you other add-ons, usages or projects in mind? If yes, would you mind to share it with us? It's of course not mandatory; just my curiosity...

**Comment by nvdaes** [comment]
Just to say that I'plan to use the keyboard extension point in the pcKbBrl add-on. Thanks for this PR. Hope this is merged when possible to test it.

File: source/inputCore.py (hunk 8 above)
Comment (seanbudd) [reply source/inputCore.py:455]: Could you provide an example of what that would look like?

File: source/inputCore.py (hunk 2 above)
Comment (ctoth) [reply source/inputCore.py:455]: I thought about this, but the actual functionality lives in the keyboard handler, it's just exposed in the input core for consistency. The keyboard handler doesn't currently seem to have any test suite I can find though (I could totally be overlooking something obvious :) )

We can and should add tests for the keyboard handler if there aren't any, but I don't necessarily think they should go in this PR.

Comment (seanbudd) [reply source/inputCore.py:455]: Thanks, I agree

**Comment by ctoth** [comment]
@seanbudd  Re #14520 we still need one more extension point for speech either directly inside or directly before `speech.manager.speak`

Unfortunately existing Remote monkeypatches `speech.manager.speak` rather than `speech.speak` (for which we could trivially use pre_speech) and to maintain backwards compatibility with existing Remote clients we'll need one more there
Still, very close now. #17429 is also almost the functionality we need, though I don't know if it quite does the job in its current form.

**Review by seanbudd — APPROVED** [review]

**Pull request merged**